src/envoy/http/authn/http_filter.cc

```diff
@@ -30,59 +57,207 @@ void AuthenticationFilter::onDestroy() {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
 }
 
-FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
+                                                        bool) {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
+  state_ = IstioAuthN::State::PROCESSING;
+  if (config_.peers_size() == 0) {
+    ENVOY_LOG(debug, "No method defined. Skip source authentication.");
+    onAuthenticatePeerDone(&headers, 0, nullptr, true);
+  } else {
+    authenticatePeer(
+        headers, config_.peers(0),
```

Comment:
can we create an object for each peer?  It has following benefits:
1)  it can carry "headers" so it doesn't need to be passed in many function arguements
2) these objects  maybe able to process con-currently.

Reply:
I'm not sure if we want to process concurrently, as we want to accept only one 'method' (running concurrent may cost more CPU if the first method to try already success).

A better implementation for this maybe pre-select the authN method to try (e.g checking if token/certificate is available for each method, then select the first one available). However, the gain may not be significant, as authenticationPeer basically doing that and return quickly if token is not available. I would like to defer that optimization until later.